apps/sms/services/js/moz_mobile_message/moz_mobile_message_shim.js

```diff
@@ -0,0 +1,210 @@
+  },
+
+  onDeleted: function ms_onDeleted(e) {
+    if (e.deletedThreadIds && e.deletedThreadIds.length) {
+      service.broadcast('threads-deleted', {
+        ids: e.deletedThreadIds
+      });
+    }
+  },
+
+  /* Method */  
+
+  getMessage: function ms_getMessag(id) {
+    return mozMobileMessage.getMessage(id);
+  },
+
+  retrieveMMS: function ms_retrieveMMS(id) {
+    return mozMobileMessage.retrieveMMS(id);
+  },
+
+  send: function ms_send(recipient, content, sendOpts) {
+    return mozMobileMessage.send(recipient, content, sendOpts);
+  },
+
+  sendMMS: function ms_sendMMS(dataOpts, sendOpts) {
+    return mozMobileMessage.sendMMS(dataOpts, sendOpts);
+  },
+
+  delete: function ms_delete(id) {
+    return mozMobileMessage.delete(id);
+  },
+
+  markMessageRead: function ms_markMessageRead(id, isRead, returnReq) {
+    return mozMobileMessage.markMessageRead(id, isRead, returnReq);
+  },
+
+  getSegmentInfoForText: function ms_getSegmentInfoForText(text) {
+    return mozMobileMessage.getSegmentInfoForText(text);
+  },
+
+  /* Stream */
+
+  /**
+   * Stream wrapper for getThreads API for returning threads in chunck.
+   * @param {*} stream channel for returning thread and flag.
+   */
+
+  getThreads: function ms_getThreads(stream) {
+    var cursor = null;
+
+    // WORKAROUND for bug 958738. We can remove 'try\catch' block once this bug
+    // is resolved
+    try {
+      cursor = mozMobileMessage.getThreads();
+    } catch(e) {
+      console.error('Error occurred while retrieving threads: ' + e.name);
+      stream.abort();
+      return;
+    }
+
+    cursor.onsuccess = function onsuccess() {
+      var result = this.result;
+
+      if (result) {
+        stream.write(result);
+        this.continue();
+        return;
+      }
+
+      stream.close();
+    };
+
+    cursor.onerror = function onerror() {
+      console.error('Reading the database. Error: ' + this.error.name);
+      stream.abort();
+    };
+  },
+
+  /**
+   * Stream wrapper for getMessages API for returning messagess in chunck.
+   * @param {*} stream channel for returning message and flag.
+   * @param {*} options Options for getMessages API.
+   */
+
+  getMessages: function ms_getMgs(stream, options) {
+    /*
+    options { 
+      filter: a MobileMessageFilter or similar object
+      invert: option to invert the selection
+    }
+
+     */
+    var invert = options.invert;
+    var filter = options.filter;
+    var cursor = mozMobileMessage.getMessages(filter, !invert);
+
+    cursor.onsuccess = function onsuccess() {
+      if (!this.done) {
+        stream.write(this.result);
+        this.continue();
+      } else {
+        stream.close();
+      }
+    };
+    cursor.onerror = function onerror() {
+      var msg = 'Reading the database. Error: ' + this.error.name;
+      console.error(msg);
+      stream.abort();
+    };
+
+    stream.cancel = function() {
+      stream.close();
+    };
+  }
+};
+
+MozMobileMessageShim.init();
```

Comment:
hmm, but since it might live in iframe temporary, where can we init if the shim is not hosted in main context nor worker? Or we should call init via bridge service?